### PR TITLE
Refactor wellstate init

### PR DESCRIPTION
**Pull request by joakim-hove**
This is the refactor part of: #3647

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Comments to this?

File: opm/simulators/wells/WellState.cpp
Comment (bska) [inline opm/simulators/wells/WellState.cpp:107]: This condition will never trigger in the current implementation since it ultimately [reduces](https://github.com/OPM/opm-common/blob/dd54d2aad1eb5517a6c4aecbedf6ecc339bfa4d9/src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp#L742-L748) to
```C++
WellType::producer() && !WellType::producer()
```
given that `WellType::injector()` is [`! WellType::producer()`](https://github.com/OPM/opm-common/blob/dd54d2aad1eb5517a6c4aecbedf6ecc339bfa4d9/src/opm/parser/eclipse/EclipseState/Schedule/ScheduleTypes.cpp#L167-L173).  Unless we consider introducing new well types (e.g., "observer"), then I suggest to remove that check.

File: opm/simulators/wells/SingleWellState.hpp
Comment (bska) [inline opm/simulators/wells/SingleWellState.hpp:73]: Tiny nit: The documented parameter name should match the actual parameter name.

**Review by bska — COMMENTED** [review]
Looks good for the most part.  Just a couple of minor points that I'd like to address before we merge this.

File: opm/simulators/wells/WellState.cpp
Comment (joakim-hove) [reply opm/simulators/wells/WellState.cpp:107]: I fully agree. The current logic was just to reproduce the existing `assert(...)` based approach (might have got that wrong though?).

File: opm/simulators/wells/SingleWellState.hpp
Comment (joakim-hove) [reply opm/simulators/wells/SingleWellState.hpp]: OK

File: opm/simulators/wells/WellState.cpp
Comment (joakim-hove) [reply opm/simulators/wells/WellState.cpp:107]: I'll push an update without the check.

**Comment by joakim-hove** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks for the updates. This looks good to me now and I'll merge once the build check completes.

**Comment by bska** [comment]
Build check is green.  I'll merge into master.

**Pull request merged**
